gather-scheduler.js

```diff
@@ -18,99 +18,114 @@
 
 class GatherScheduler {
 
-  static _runPhase(gatherers, gatherFun) {
-    return gatherers.reduce(
-      (chain, gatherer) => chain.then(_ => gatherFun(gatherer)),
-      Promise.resolve()
-    );
+  static runSeries(fns) {
+    const args = Array.from(arguments).slice(1);
+
+    return fns.reduce((prev, curr) => {
+      return prev.then(_ => curr.apply(null, args));
+    }, Promise.resolve());
   }
 
-  static run(gatherers, options) {
-    const driver = options.driver;
-    const url = options.url;
+  static loadPage(driver, gatherers, options) {
     const loadPage = options.flags.loadPage;
-    const emulateMobileDevice = options.flags.mobile;
-    const tracingData = {};
-    const artifacts = [];
+    const url = options.url;
 
-    if (url === undefined || url === null) {
-      throw new Error('You must provide a url to scheduler');
+    if (loadPage) {
+      return driver.gotoURL(url, driver.WAIT_FOR_LOADED);
     }
 
-    return driver.connect()
-
-      // Enable emulation.
-      .then(_ => {
-        if (emulateMobileDevice) {
-          return driver.beginEmulation();
-        }
-
-        return Promise.resolve();
-      })
+    return Promise.resolve();
+  }
 
-      // Clean all browser caches.
-      .then(_ => driver.cleanAndDisableBrowserCaches())
+  static reloadPage(driver, gatherers, options) {
+    // Such a hack... since a Page.reload command does not let
+    // a service worker take over we have to trick the browser into going away
+    // and then coming back.
+    return driver.sendCommand('Page.navigate', {url: 'about:blank'}).then(_ => {
+      return driver.gotoURL(options.url, driver.WAIT_FOR_LOADED);
+    });
+  }
 
+  static setupDriver(driver, gatherers, options) {
+    return new Promise((res, _) => {
```

Comment:
since there is some logic inside of the block here I wanted to wrap it all in a promise constructor to handle the case where something could potentially throw 💥